### ISSUE TITLE
memory leak with H2O's DataTable

### ISSUE THREAD

**Issue by pseudotensor** [Microsoft/LightGBM#1968]
[creep.zip](https://github.com/Microsoft/LightGBM/files/2794219/creep.zip)

When doing a permutation variable importance one loops over many shuffled columns and get predictions over and over.

We have seen that LightGBM has a memory leak when performing this operation.  XGBoost has no memory leak.

Just run the script and keep track of (say) resident memory for the process.  

It starts out bouncing around 0.4GB, but by n=1300 columns later bounces around 1.0GB.  Once it hit n=2100, it's bouncing around 1.6GB.  By n=2500 columns, memory bouncing around 1.8GB.

You can change get_new_preds to False to just see how datatable is not using any extra memory.  Go back and then try xgboost and no such memory leak occurs.

Here is the location of the data, and one needs to fix the path used in the code: https://www.kaggle.com/c/santander-value-prediction-challenge/download/train.csv (you'll need a kaggle account).

Thanks!
Jon

**Comment by pseudotensor** [comment]
Do you expect a reference cycle to exist inside lightgbm so gc.collect() is required?  Otherwise, reference counting is enough.

**Comment by guolinke** [comment]
I am not familiar with python. a reference: https://stackoverflow.com/questions/15455048/releasing-memory-in-python

**Comment by pseudotensor** [comment]
Python reference counting should be enough here, so the issue is that you probably don’t free memory in C, e.g. in the destructor.

**Comment by guolinke** [comment]
@pseudotensor 
We use all RAII style in cpp side, there are not raw pointers.

**Comment by guolinke** [comment]
@pseudotensor 
I write a minimal test, and didn't spot the memory leak:
```
import numpy as np
import lightgbm as lgb

nrow = 100000
ncol = 100
X = np.random.randn(nrow, ncol)
y = np.random.randn(nrow)
model = lgb.LGBMRegressor(objective='regression', n_jobs=20)  # 40 very slow
model.fit(X, y)

for n in range(10000):
    X = np.random.randn(nrow // 100, ncol)
    new_preds = model.predict(X)
```

**Comment by pseudotensor** [comment]
Did you try my test to confirm a problem with LightGBM and no problem with XGBoost?  Good to bisect since any minimal test may miss something.

**Comment by pseudotensor** [comment]
Issue was with datatable -> lightgbm leads lightgbm to use numpy buffer capability, and that keeps the buffer around.  Solution is to avoid using the buffer facility and create a numpy/pandas object before passing into lgbm.

XGBoost didn't have this problem because it handles datatable directly, which is something LightGBM could do.

**Comment by pseudotensor** [comment]
I take back closing this.  I think lightgbm must not be freeing the buffer facility, which direct datatable -> lightgbm ingestion would rely upon.

https://docs.python.org/3/c-api/buffer.html .

That is, it's clearly a lightgbm bug, it's not datatables job to manage the conversion from the buffer faciltiy:

```
In both cases, PyBuffer_Release() must be called when the buffer isn’t needed anymore. Failure to do so could lead to various issues such as resource leaks.
```

**Comment by guolinke** [comment]
I see.
I think the best solution is supporting the datatable directly, without introducing these buffer issue.

**Comment by guolinke** [comment]
@pseudotensor Thanks! That is very welcome!
I make a temporary fix for this, but directly convert it to numpy.
